### DISCUSSION TITLE
Usage of error stream

### DISCUSSION THREAD

**Discussion by pgschulz** [aws/aws-sam-cli#7065] [Questions] [answered]
Sam prints a lot of information to the error stream during build and deploy operations. Is this behavior expected or a bug?

**Comment by uldyssian-sh** [comment, marked as answer]
Yes, this behavior is expected.

SAM intentionally writes a lot of its diagnostic, progress, and informational output to **stderr**, not stdout. This includes messages printed during `sam build`, `sam deploy`, and related commands.

Why SAM does this:

- `stdout` is reserved for machine-readable output (JSON, templates, values that users or scripts may want to capture).
- `stderr` is used for:
  - progress logs
  - status messages
  - warnings
  - debug information
  - non-fatal errors

This design allows commands like:

sam deploy --guided --output json > output.json

to work reliably without polluting stdout with human-readable logs.

Is it a bug?

No. It is a deliberate CLI design choice, and SAM follows the same pattern as many other developer tools (including parts of the AWS CLI itself).

Implications for scripting and CI:

- If you are piping or redirecting output, expect most SAM logs to appear on stderr.
- To suppress noise in CI logs, you can redirect stderr separately, for example:

sam build 2>/dev/null

- To capture everything:

sam deploy > all.log 2>&1

Summary:

SAM printing lots of information to stderr is expected behavior.
stderr is used for logs and progress; stdout is kept clean for structured output.
This is by design, not a bug.

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
